MapleServer2/PacketHandlers/Game/WardrobeHandler.cs

```diff
@@ -0,0 +1,147 @@
+                break;
+            default:
+                LogUnknownMode(mode);
+                break;
+        }
+    }
+
+    private static void HandleSave(GameSession session, PacketReader packet)
+    {
+        int index = packet.ReadInt();
+        int type = packet.ReadInt();
+        Dictionary<ItemSlot, Item> equips = type == 0 ? session.Player.Inventory.Equips : session.Player.Inventory.Cosmetics;
+
+        if (!session.Player.TryGetWardrobe(index, out Wardrobe wardrobe))
+        {
+            return;
+        }
+
+        foreach (Item equip in equips.Values)
+        {
+            // remove to not save Hair, Face, Ears, or Face Decoration
+            if (equip.ItemSlot is ItemSlot.HR or ItemSlot.ER or ItemSlot.FD or ItemSlot.FA)
+            {
+                equips.Remove(equip.ItemSlot);
+                continue;
+            }
+        }
+
+        wardrobe.Equips = new(equips);
+        wardrobe.Type = type;
+
+        DatabaseManager.Wardrobes.Update(wardrobe, session.Player.CharacterId);
+        session.Send(WardrobePacket.Load(wardrobe));
+    }
+
+    private static void HandleEquip(GameSession session, PacketReader packet)
+    {
+        int index = packet.ReadInt();
+        int type = packet.ReadInt();
+
+        if (!session.Player.TryGetWardrobe(index, out Wardrobe wardrobe))
+        {
+            return;
+        }
+
+        int failEquipCount = 0;
+        foreach (KeyValuePair<ItemSlot, Item> item in wardrobe.Equips)
+        {
+            if (!session.Player.Inventory.Equip(session, item.Value.Uid, item.Key))
+            {
+                failEquipCount++;
+            }
+        }
+
+        SystemNotice notice = failEquipCount == 0 ? SystemNotice.ClosetMsgSuccess : SystemNotice.ClosetMsgItemNotExist;
+        session.Send(NoticePacket.Notice(notice, NoticeType.Chat | NoticeType.FastText));
+    }
+
+    private static void HandleReset(GameSession session, PacketReader packet)
+    {
+        int index = packet.ReadInt();
+        if (!session.Player.TryGetWardrobe(index, out Wardrobe wardrobe))
+        {
+            return;
+        }
+
+        wardrobe.Equips.Clear();
+
+        DatabaseManager.Wardrobes.Update(wardrobe, session.Player.CharacterId);
+        session.Send(WardrobePacket.Load(wardrobe));
+    }
+
+    private static void HandleSetShortcut(GameSession session, PacketReader packet)
+    {
+        int index = packet.ReadInt();
+        int keyCode = packet.ReadInt();
+
+        if (!session.Player.TryGetWardrobe(index, out Wardrobe wardrobe))
+        {
+            return;
+        }
+
+        wardrobe.Key = keyCode;
+    }
+
+    private static void HandleRename(GameSession session, PacketReader packet)
+    {
+        int index = packet.ReadInt();
+        string name = packet.ReadUnicodeString();
+
+        if (!session.Player.TryGetWardrobe(index, out Wardrobe wardrobe))
+        {
+            _ = new Wardrobe(0, 0, index, name, new(), session.Player);
```

Comment:
I think there should be an explicit function call to assign the wardrobe to the player, instead of using the constructor for that.
I would prefer these two function calls instead of this line.
or a function that calls these two, either would be good I think.

       DatabaseManager.Wardrobes.Insert(this, player.CharacterId);
        player.Wardrobes.Insert(index, this);

I dont like calling the Wardrobe constructor for that.

Reply:
I think the database insert inside the constructor it makes sense but it should add to player while creating like player.Wardrobe.Insert(new Wardrobe())